core/src/main/java/ai/timefold/solver/core/impl/domain/variable/declarative/DefaultShadowVariableSessionFactory.java

```diff
@@ -0,0 +1,185 @@
+    }
+
+    static <Solution_> void visitGraph(
+            SolutionDescriptor<Solution_> solutionDescriptor,
+            VariableReferenceGraph<Solution_> variableReferenceGraph, Object[] entities,
+            IntFunction<TopologicalOrderGraph> graphCreator) {
+        var declarativeShadowVariableDescriptors = solutionDescriptor.getDeclarativeShadowVariableDescriptors();
+        var variableIdToUpdater = new HashMap<VariableId, VariableUpdaterInfo>();
+
+        // Maps a variable id to it source aliases;
+        // For instance, "previousVisit.startTime" is a source alias of "startTime"
+        // One way to view this concept is "previousVisit.startTime" is a pointer
+        // to "startTime" of some visit, and thus alias it.
+        Map<VariableId, Set<VariableSourceReference>> declarativeShadowVariableToAliasMap = new HashMap<>();
+
+        // Create graph node for each entity/declarative shadow variable pair
+        for (var entity : entities) {
+            for (var declarativeShadowVariableDescriptor : declarativeShadowVariableDescriptors) {
+                var entityClass = declarativeShadowVariableDescriptor.getEntityDescriptor().getEntityClass();
+                if (entityClass.isInstance(entity)) {
+                    var variableId = new VariableId(entityClass, declarativeShadowVariableDescriptor.getVariableName());
+                    var updater = variableIdToUpdater.computeIfAbsent(variableId, ignored -> new VariableUpdaterInfo(
+                            entityClass,
+                            declarativeShadowVariableDescriptor.getVariableName(),
+                            declarativeShadowVariableDescriptor,
+                            declarativeShadowVariableDescriptor.getEntityDescriptor().getInvalidityMarkerVariableDescriptor(),
+                            declarativeShadowVariableDescriptor.getMemberAccessor(),
+                            declarativeShadowVariableDescriptor.getCalculator()::executeGetter));
+                    variableReferenceGraph.addVariableReferenceEntity(
+                            variableId,
+                            entity,
+                            updater);
+                    for (var sourceRoot : declarativeShadowVariableDescriptor.getSources()) {
+                        for (var source : sourceRoot.variableSourceReferences()) {
+                            if (source.downstreamDeclarativeVariable() != null) {
+                                declarativeShadowVariableToAliasMap
+                                        .computeIfAbsent(source.downstreamDeclarativeVariable(),
+                                                ignored -> new LinkedHashSet<>())
+                                        .add(source);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        // Create variable listeners for each declarative shadow variable descriptor
```

Comment:
Where exactly are the listeners being created? I don't see any use of `VariableListener` here. Maybe let's call it something else, to not mix concepts?